### PR TITLE
Support Watir-Nokogiri

### PR DESCRIPTION
**Pull request by pvmeerbe**
Pull request to also support : https://github.com/jkotests/watir-nokogiri

__Goal:__
Given a rails project with integration tests build around page-objects. 
Now we also want to write functional tests on the controller and quickly test the @response.body of the ActionController::TestCase (i.e. rendered HTML of the view)  for the presence of some elements using the mapped elements of the page objects (i.e. avoid repeating the xpaths here). 

__Issue:__
For these functional tests  selenium-webdriver or watir-webdriver cannot be used directly as no rails app is running. We do not want to run one as the requests will be considerably slower then the current approach with ActionController::TestCase (Controller is instantiated directly)

__Solution:__
A gem exists, called watir-nokogiri, which allows the use of the watir API on a chunk of HTML. 
In other words the WatirNokogiri::Document object can be used instead of a  Watir::Browser object to verify whether the mapped elements of the page object are present in the HTML

This is a similar solution as the one Capybara provides : http://robots.thoughtbot.com/use-capybara-on-any-html-fragment-or-page

Note: Not all the Watir API commands are supported, only the ones for identifying/reading elements on the HTML page. An error  "Not implemented" is raised for the other methods

__Example:__
browser= WatirNokogiri::Document.start('my_page.html')
my_page_object = MyPageObject.new(browser)

allows to use:
on_page(MyPageObject).my_mapped_field_element.exists? 
or even:
on_page(MyPageObject).my_mapped_field?

### CODE REVIEW

**Comment by daneandersen** [comment]
Can watir nokogiri completely replace watir browser? Or do you have to mix and match? 

If you could use just watir nokogiri, this might make a good platform

**Comment by daneandersen** [comment]
I looked over the code, and it looks like WatirNokogiri can't completely replace WatirWebdriver. WatirNokogiri doesn't implement any interaction methods, so You would have to set up two different browsers and instantiate page objects manually (instead of using the factory methods)

Overall this seems like a difficult and non-optimal workflow for the users of page-object. I would be more interested in seeing a platform that can intelligently choose WatirWebdriver or WatirNokogiri based on performance. That way, you would set it up once, and use it throughout instead of having to pick and choose where you need to use what. Another benefit of this is that if WatirNokogiri implemented any other WatirWebdriver behaviors, you could change it in the platform, and everyone using it would immediately get the benefit instead of having to adapt their code.

**Comment by jkotests** [comment]
As author of the Watir-Nokogiri gem, it would be great to see more usage of the gem. However, I do not see a lot of benefit of making this change to the Page-Object gem. The user base for the Watir-Nokogiri gem is pretty small. If adding the support is really just those few lines of code, I think it would be reasonable to suggest Watir-Nokogiri users monkey patch their Page-Object gem.

**Comment by pvmeerbe** [comment]
>You would have to set up two different browsers and instantiate page objects manually (instead of using >the factory methods)

I was more thinking of  creating a method which:
- sets the @browser variable  to  WatirNokogiri::Document.start(@response.body)
- is automatically called whenever a PageObject is needed in the functional tests

This way you can still use the PageFactory methods without the user  needing to create/maintain this
He can directly use 'on_page(MyPageObject)'

> I would be more interested in seeing a platform that can intelligently choose WatirWebdriver or >WatirNokogiri based on performance.'

The problem is not only performance related. in the functional test there is no rails server running so the  watir http requests cannot be handled. Launching a server in this case will then seriously affect performance in this case. So somehow the platform should be able to intelligently detect that a server is not available. This would mean that the platform needs to know on which ip/port combination the server is running + some kind of timeout needs to be configured to prevent glitches from automatically switch between drivers. This automatically switching could IMHO also cause unwanted behavior during integration testing as the watir-nokogiri gem doesn't implement any interaction methods : imagine that for some reason the timeout is encountered  and the driver switches from webdriver to watir-nokogiri...


@jkotests : thanks for the gem and all the work you put it in ! As you suggested I currently monkey patched the page-object gem. However I do not really like monkey patching, hence the small request ;)

**Comment by anilreddy** [comment]
+1 it makes tests run faster

**Comment by daneandersen** [comment]
If monkey-patching, as @jkotests suggested, is undesirable, I would suggest seeing if you can create a new platform. page-object has a method of registering a new platform. 

```ruby
PageObject::Platforms.register(:watir_nokogiri, PageObject::Platforms::WatirNokogiri)
```

You could bundle it all up as a gem so that when someone requires it, it adds the platform to their project. You would have to write some code for how the platform works, but it would mainly be lifting the watir-webdriver platform and adding the changes you submitted to that.

**Comment by daneandersen** [comment]
I'm going to close this pull request since this doesn't seem like the right approach. Feel free to reopen if you feel it needs further discussion.
